Command to Deprecate Analysis Projects

jasonwalker80 opened the pull request:
New analysis project status Depcrecated and a new Deprecate command. Support for Deprecated status added to existing status update commands.

tmooney commented on lib/perl/Genome/Config/AnalysisProject/Command/Deprecate.pm:
```diff
@@ -0,0 +1,55 @@
+
+sub help_detail {
+    return <<"EOS"
+Given some analysis projects, this will deprecate the analysis project.  This includes setting all config profile items to disabled and updating the project status to Deprecated.
```
There is a disagreement in the number of analysis projects in the first sentence.

ebelter commented:
+1

tmooney commented on lib/perl/Genome/Config/AnalysisProject/Command/Release.pm:
```diff
@@ -46,7 +46,7 @@ sub __errors__ {
     my @errors = $self->SUPER::__errors__(@_);
     for my $status (map{$_->status} $self->analysis_projects){
-        unless(grep{$_ eq $status} ("Pending", "Hold", "In Progress")){
+        unless(grep{$_ eq $status} ("Pending", "Hold", "In Progress","Deprecated")){
```
To verify: we want to **allow** setting the status back to `Hold` or `In Progress` from `Deprecated`?

Either way, we may want to put a similar status check on the new command to prevent deprecating AnPs that are in the various "final" statuses (like `Purged` or `Template`).

jasonwalker80 commented:
Good catch Tom, I removed Deprecated from the list of allowed statuses for Hold and Release.  Deprecated should be a final status along with Archived, Purged, Template, etc.

susannasiebert commented:
The capitalization of the `valid_values` for `status` is inconsistent between `Genome::Config::AnalysisProject` and `Genome::Config::Profile::Item`. It seems like in general we tend to use all-lowercase in our code base. +1 either way.

tmooney commented:
I'd still like to see this command check the status of the AnPs it's operating on to make sure they're not already `Purged` or other statuses that wouldn't make sense to deprecate.

jasonwalker80 commented:
Updated version.  I modified the command to use Base class to reuse the status validation logic.

@susannasiebert you are correct about the inconsistency in capitalization.

tmooney commented on lib/perl/Genome/Config/AnalysisProject/Command/Deprecate.t:
```diff
@@ -0,0 +1,40 @@
+my $cmd = $class->create(analysis_project => $ap);
+ok($cmd->execute(), 'succesful command execution');
+
+is($ap->status, 'Deprecated', 'it should set the status to hold');
```
Perhaps `to Deprecated` instead of `to hold`.

jasonwalker80 replied:
Correction made.  copy/paste error

tmooney commented:
Since #807 has been merged, we'll want to add `Deprecated` to the excluded list for `Genome::Config::AnalysisProject->is_current`.  (I can do this as a separate PR if you like.)

+1 aside from that and my nitpicking of test descriptions!

jasonwalker80 commented:
I can't seem to find the code for `Genome::Config::AnalysisProject->is_current`.  That will have to make it in as a separate PR.  Merging after tests pass.
